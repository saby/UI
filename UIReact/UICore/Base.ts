export { default as Control}  from './_base/Control';

export {IControlOptions, TemplateFunction} from './_base/interfaces';
