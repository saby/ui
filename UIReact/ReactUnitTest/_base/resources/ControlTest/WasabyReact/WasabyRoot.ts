import { Control, TemplateFunction } from 'UI/Base';

// @ts-ignore
import template = require('wml!ReactUnitTest/_base/resources/ControlTest/WasabyReact/WasabyRoot');

export default class WasabyRoot extends Control {
    readonly _template: TemplateFunction = template;
}
