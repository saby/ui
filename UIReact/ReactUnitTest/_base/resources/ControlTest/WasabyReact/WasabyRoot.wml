<div>
    <ws:partial template="{{_options.control}}" />
</div>
