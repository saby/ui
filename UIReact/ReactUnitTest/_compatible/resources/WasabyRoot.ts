import { Control, TemplateFunction } from 'UI/Base';

// @ts-ignore
import template = require('wml!ReactUnitTest/_compatible/resources/WasabyRoot');

// Синхронная загрузка, чтобы замокать асинхронный load.
import 'ReactUnitTest/_compatible/resources/WS3Control';

export default class WasabyRoot extends Control {
    readonly _template: TemplateFunction = template;
}
