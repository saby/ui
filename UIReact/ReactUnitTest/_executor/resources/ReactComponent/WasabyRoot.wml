<div>
    <ws:if data="{{_options.reactFunction}}">
        <ReactUnitTest._executor.resources.ReactComponent.ReactFunction />
    </ws:if>
    <ws:if data="{{_options.reactComponent}}">
        <ReactUnitTest._executor.resources.ReactComponent.ReactComponent />
    </ws:if>
    <ws:if data="{{_options.extendedReactComponent}}">
        <ReactUnitTest._executor.resources.ReactComponent.ExtendedReactComponent />
    </ws:if>
    <ws:if data="{{_options.memoizedReactComponent}}">
        <ReactUnitTest._executor.resources.ReactComponent.MemoizedReactComponent forwardedRef="{{ htmlElementRef }}" />
    </ws:if>
    <ws:if data="{{_options.reactPureComponent}}">
        <ReactUnitTest._executor.resources.ReactComponent.ReactPureComponent />
    </ws:if>
    <ws:if data="{{_options.wasabyControl}}">
        <ReactUnitTest._executor.resources.ReactComponent.WasabyInner />
    </ws:if>
    <ws:if data="{{_options.wasabyControlContent}}">
        <ReactUnitTest._executor.resources.ReactComponent.WasabyInnerContent />
    </ws:if>
    <ws:if data="{{_options.withForwardRef}}">
        <ReactUnitTest._executor.resources.ReactComponent.WithForwardRef forwardedRef="{{ htmlElementRef }}" />
    </ws:if>
    <ws:if data="{{_options.contextProvider}}">
        <ReactUnitTest._executor.resources.ReactComponent.ContextProvider value="valueFromWasabyRoot">
            <ReactUnitTest._executor.resources.ReactComponent.ContextUser />
        </ReactUnitTest._executor.resources.ReactComponent.ContextProvider>
    </ws:if>
    <ws:if data="{{_options.contextConsumer}}">
        <ReactUnitTest._executor.resources.ReactComponent.ContextConsumer>
            <ws:children>
                <ws:Function>
                    ReactUnitTest/_executor/resources/ReactComponent/contextConsumerChildren:default
                </ws:Function>
            </ws:children>
        </ReactUnitTest._executor.resources.ReactComponent.ContextConsumer>
    </ws:if>
</div>
