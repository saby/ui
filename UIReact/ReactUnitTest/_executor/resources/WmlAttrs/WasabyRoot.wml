<div>
    <ws:if data="{{_options.dataQaFn}}">
        <ReactUnitTest._executor.resources.WmlAttrs.ReactFunction attr:data-qa="data-qa-react-function" />
    </ws:if>
    <ws:if data="{{_options.dataQaCls}}">
        <ReactUnitTest._executor.resources.WmlAttrs.ReactComponent attr:data-qa="data-qa-react-component" />
    </ws:if>
</div>
