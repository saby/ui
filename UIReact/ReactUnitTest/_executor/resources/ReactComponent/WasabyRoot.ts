import { Control, TemplateFunction, IControlOptions } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_executor/resources/ReactComponent/WasabyRoot';

interface IWasabyRootOptions extends IControlOptions {
    reactFunction?: boolean;
    reactComponent?: boolean;
    extendedReactComponent?: boolean;
    memoizedReactComponent?: boolean;
    reactPureComponent?: boolean;
    wasabyControl?: boolean;
    wasabyControlContent?: boolean;
    withForwardRef?: boolean;
}
export default class WasabyRoot extends Control<IWasabyRootOptions> {
    _template: TemplateFunction = template;

    // Для проверки, правильно ли пролетает forwardedRef через генератор.
    // В классовом компоненте он должен остаться forwardedRef, в forwardRef - преобразоваться в ref.
    htmlElementRef(element?: HTMLElement): void {
        element.className = 'patchedFromForwardedRef';
    }
}
