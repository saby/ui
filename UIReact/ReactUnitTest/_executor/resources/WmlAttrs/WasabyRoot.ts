import { Control, TemplateFunction, IControlOptions } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_executor/resources/WmlAttrs/WasabyRoot';

interface IWasabyRootOptions extends IControlOptions {
    reactFunction?: boolean;
    reactComponent?: boolean;
    dataQaFn?: boolean;
    dataQaCls?: boolean;
    clearProps?: boolean;
}
export default class WasabyRoot extends Control<IWasabyRootOptions> {
    _template: TemplateFunction = template;
}
