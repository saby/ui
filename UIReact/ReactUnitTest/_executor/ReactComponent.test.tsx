/**
 * @jest-environment jsdom
 */
import { act } from 'react-dom/test-utils';
import { Control } from 'UICore/Base';

import WasabyRoot from './resources/ReactComponent/WasabyRoot';

const creator = Control.createControl;

describe('Определение типа компонента', () => {
    let container: HTMLDivElement;

    /**
     * Эта функция существует для того, чтобы не забывать оборачивать тики в act.
     * Это нужно, чтобы реакт реагировал на изменение времени и обновлял компоненты.
     * @param duration Значение, на которое должно продвинуться время.
     */
    function tick(duration: number): void {
        act(() => {
            jest.advanceTimersByTime(duration);
        });
    }

    beforeEach(() => {
        /*
        _afterMount и _afterUpdate зовутся в отдельной таске, чтобы браузер мог отрисовать кадр.
        Чтобы не делать тесты асинхронными, мы просто мокнем таймеры и сами будем управлять временем.
         */
        jest.useFakeTimers();

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    afterEach(() => {
        jest.useRealTimers();
        container.remove();
    });

    it('Functional Component', async () => {
        // region Setup
        act(() => {
            creator(WasabyRoot, { reactFunction: true }, container);
        });
        tick(0);
        // endregion

        expect(container).toMatchSnapshot();
    });

    it('React Component', async () => {
        // region Setup
        act(() => {
            creator(WasabyRoot, { reactComponent: true }, container);
        });
        tick(0);
        // endregion

        expect(container).toMatchSnapshot();
    });

    it('React Component, наследник другого React Component', async () => {
        // region Setup
        act(() => {
            creator(WasabyRoot, { extendedReactComponent: true }, container);
        });
        tick(0);
        // endregion

        expect(container).toMatchSnapshot();
    });

    it('React Component, обёрнутый в memo', async () => {
        // region Setup
        act(() => {
            creator(WasabyRoot, { memoizedReactComponent: true }, container);
        });
        tick(0);
        // endregion

        expect(container).toMatchSnapshot();
    });

    it('React PureComponent', async () => {
        // region Setup
        act(() => {
            creator(WasabyRoot, { reactPureComponent: true }, container);
        });
        tick(0);
        // endregion

        expect(container).toMatchSnapshot();
    });

    it('Wasaby Controls', async () => {
        // region Setup
        act(() => {
            creator(WasabyRoot, { wasabyControl: true }, container);
        });
        tick(0);
        // endregion

        expect(container).toMatchSnapshot();
    });

    it('Wasaby Controls with content', async () => {
        // region Setup
        act(() => {
            creator(WasabyRoot, { wasabyControlContent: true }, container);
        });
        tick(0);
        // endregion

        expect(container).toMatchSnapshot();
    });

    it('Вставка forwardRef в wasaby', async () => {
        // region Setup
        act(() => {
            creator(WasabyRoot, { withForwardRef: true }, container);
        });
        tick(0);
        // endregion

        expect(container).toMatchSnapshot();
    });
});
