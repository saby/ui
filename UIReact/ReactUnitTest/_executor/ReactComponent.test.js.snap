// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Определение типа компонента Functional Component 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div>
      Is React Function
    </div>
  </div>
</div>
`;

exports[`Определение типа компонента React Component 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div>
      Is React Component
    </div>
  </div>
</div>
`;

exports[`Определение типа компонента React Component, наследник другого React Component 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div>
      Is Extended React Component
    </div>
  </div>
</div>
`;

exports[`Определение типа компонента React Component, обёрнутый в memo 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      class="patchedFromForwardedRef"
    />
  </div>
</div>
`;

exports[`Определение типа компонента React PureComponent 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div>
      Is React Pure Component
    </div>
  </div>
</div>
`;

exports[`Определение типа компонента Wasaby Controls 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      tabindex="0"
    >
      Is Wasaby Control
    </div>
  </div>
</div>
`;

exports[`Определение типа компонента Wasaby Controls with content 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      tabindex="0"
    >
      <div
        tabindex="0"
      >
        Is content Template
      </div>
    </div>
  </div>
</div>
`;

exports[`Определение типа компонента Вставка forwardRef в wasaby 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      class="patchedFromForwardedRef"
    >
      <button>
        0
      </button>
    </div>
  </div>
</div>
`;
